optimum/habana/distributed/fastddp.py

```diff
@@ -0,0 +1,135 @@
+        fusion_buffer = torch.zeros(size=(grad_elem_count,), dtype=fusion_buffer_dtype, device='hpu:0')
+        model.register_buffer(_ALL_REDUCE_FUSION_BUFFER_NAME, fusion_buffer, persistent=False)
+
+        # Build the fusion information necessary for packing and unpacking processes.
+        grad_elem_count = 0
+        fusion_entries = []
+        for param in model.parameters():
+            if param.grad is None:
+                continue
+            grad_numel = torch.numel(param.grad)
+            fused_view = fusion_buffer[grad_elem_count:grad_elem_count + grad_numel].reshape(param.grad.shape)
+            fusion_entries.append((param, fused_view))
+            grad_elem_count += grad_numel
+        model._all_reduce_fusion_entries = fusion_entries
+
+        # Instruct the following logic to record packing and unpacking HPU graphs based on the newly created fusion buffer.
+        if use_hpu_graph:
+            pack_graph = None
+            unpack_graph = None
+
+    assert (fusion_buffer is not None) and (fusion_entries is not None)
+
+    # Pack the gradients into the fusion buffer.
+    def pack_grads():
+        world_size_inv = 1.0 / torch.distributed.group.WORLD.size()
+        for param, fused_view in fusion_entries:
+            grad = param.grad
+            assert grad is not None
+            if grad.dtype != fusion_buffer_dtype:
+                grad = grad.to(fusion_buffer_dtype)
+            grad = grad * world_size_inv
+            fused_view.copy_(grad, non_blocking=True)
+
+    if use_hpu_graph:
+        if pack_graph is None:
+            import habana_frameworks.torch as ht
+            pack_graph = ht.hpu.HPUGraph()
+            with ht.hpu.stream(ht.hpu.Stream()):
+                pack_graph.capture_begin()
+                pack_grads()
+                pack_graph.capture_end()
+            model._all_reduce_gradient_pack_graph = pack_graph
+
+        pack_graph.replay()
+    else:
+        pack_grads()
+
+    # Invoke an all-reduce operation of the fused gradients.
+    torch.distributed.all_reduce(fusion_buffer, group=torch.distributed.group.WORLD, async_op=True)
+
+    # Unpack the gradients back to the model parameters.
+    def unpack_grads():
+        for param, fused_view in fusion_entries:
+            grad = param.grad
+            assert grad is not None
+
+            if fused_view.dtype != grad.dtype:
+                fused_view = fused_view.to(grad.dtype)
+
+            grad.copy_(fused_view, non_blocking=True)
+
+    if use_hpu_graph:
+        if unpack_graph is None:
+            import habana_frameworks.torch as ht
+            unpack_graph = ht.hpu.HPUGraph()
+            with ht.hpu.stream(ht.hpu.Stream()):
+                unpack_graph.capture_begin()
+                unpack_grads()
+                unpack_graph.capture_end()
+            model._all_reduce_gradient_unpack_graph = unpack_graph
+
+        unpack_graph.replay()
+    else:
+        unpack_grads()
```

Comment:
It may only happen when the set of "active" gradients change after first step. It's technically possible, but it's a corner edge I did not invest time addressing. In such a case, I could simply recreate the fusion information and re-record the graphs, but I doubt it's a case worth covering.